classes/village/VillageUpgradeDto.php

```diff
@@ -0,0 +1,25 @@
+<?php
+
+class VillageUpgradeDto {
+    public function __construct(
+        public ?int $id = null,
+        public string $key,
+        public int $village_id,
+        public string $status = VillageUpgradeConfig::UPGRADE_STATUS_LOCKED,
+        public ?int $research_progress = null,
+        public ?int $research_progress_required = null,
+        public ?int $research_progress_last_updated = null,
+        // display-only properties
+        public string $name = "",
+        public string $description = "",
+        public ?int $materials_research_cost = null,
+        public ?int $food_research_cost = null,
+        public ?int $wealth_research_cost = null,
+        public ?int $research_time = null,
+        public ?string $research_time_remaining = null,
+        public ?int $materials_upkeep = null,
+        public ?int $food_upkeep = null,
+        public ?int $wealth_upkeep = null,
```

Comment:
I have comments below about this but the "all these values can potentially be null/empty" is something I think we can potentially use two versions of this class for

- One completely without these fields, for the cases where they are not present
- One with these fields marked as not nullable, for the cases where they are present

The `research_time_remaining` is a bit different where it's totally valid for that variable to independently be null or not null depending on whether or not there's research happening going on. But some of these others I think are more grouped around "was config loaded or not" and that's a pattern I think having a different class for can help with. See below for more detailed suggestion